### PR TITLE
Make all modification methods respect prune_previous_versions

### PR DESCRIPTION
**Pull request by phoebusm**
#### Reference Issues/PRs

https://github.com/man-group/ArcticDB/issues/711
#### What does this implement or fix?
Add prune_previous option to APIs which still don't support it
#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: cpp/arcticdb/version/test/test_version_store.cpp [hunk 1]
```diff
@@ -194,7 +194,7 @@ TEST_F(VersionStoreTest, SortMerge) {
         test_store_->append_incomplete_frame(symbol, std::move(frame.input_frame_));
     }
 
-    test_store_->sort_merge_internal(symbol, std::nullopt, true, false, false, false);
+    test_store_->sort_merge_internal(symbol, std::nullopt, true, false, false, false, false);
```
Comment (poodlewars) [inline cpp/arcticdb/version/test/test_version_store.cpp:197]: 5 bools in a row isn't a good API (nor was 4 before your change!). Refactor by pulling out a struct for the options?

File: cpp/arcticdb/version/local_versioned_engine.hpp [hunk 2]
```diff
@@ -81,7 +81,8 @@ class LocalVersionedEngine : public VersionedEngine {
     VersionedItem delete_range_internal(
         const StreamId& stream_id,
         const UpdateQuery& query,
-        bool dynamic_schema) override;
+        bool dynamic_schema,
+        bool prune_previous_versions) override;
```
Comment (poodlewars) [inline cpp/arcticdb/version/local_versioned_engine.hpp:85]: Too many bools - pull out a struct for the options?

File: cpp/arcticdb/version/version_store_api.hpp [hunk 3]
```diff
@@ -170,7 +171,8 @@ class PythonVersionStore : public LocalVersionedEngine {
             bool append,
             bool convert_int_to_float,
             bool via_iteration,
-            bool sparsify
+            bool sparsify,
+            bool prune_previous_versions = false
```
Comment (poodlewars) [inline cpp/arcticdb/version/version_store_api.hpp:175]: Why is this the only one with a default???

File: python/tests/conftest.py [hunk 4]
```diff
@@ -238,6 +238,26 @@ def arctic_library(arctic_client, lib_name):
     return arctic_client.create_library(lib_name)
 
 
+@pytest.fixture(
+    scope="function",
+    params=[
```
Comment (poodlewars) [inline python/tests/conftest.py:243]: Why is this parameterized? Isn't just LMDB fine? Isn't there a similar fixture already?

File: python/tests/integration/arcticdb/version_store/test_deletion.py [hunk 5]
```diff
@@ -540,6 +540,37 @@ def test_delete_date_range_with_strings(version_store_factory, index_start):
     assert_frame_equal(vit.data, df)
 
 
+@pytest.mark.parametrize("prune_previous_versions", [True, False])
```
Comment (poodlewars) [inline python/tests/integration/arcticdb/version_store/test_deletion.py:543]: Most of this test is the same as the one above it?

**Review by poodlewars — CHANGES_REQUESTED** [review]

File: cpp/arcticdb/version/version_store_api.hpp (hunk 3 above)
Comment (phoebusm) [reply cpp/arcticdb/version/version_store_api.hpp:175]: Yes misread that. It is unnecessary

File: python/tests/conftest.py (hunk 4 above)
Comment (phoebusm) [reply python/tests/conftest.py:243]: I was trying to create a library fixture similar to existing `def basic_store()`
It is used by `test_write_parallel_sort_merge` which tests parallelism. So I think it's safer to test multiple storages.

File: python/tests/integration/arcticdb/version_store/test_deletion.py (hunk 5 above)
Comment (phoebusm) [reply python/tests/integration/arcticdb/version_store/test_deletion.py:543]: Yes but I don't want to make `def test_delete_date_range_with_strings` more complicated. 
And I only need to test against lmdb

File: cpp/arcticdb/version/local_versioned_engine.hpp [hunk 6]
```diff
@@ -49,6 +49,14 @@ struct KeySizesInfo {
     size_t uncompressed_size; // bytes
 };
 
+struct SortMergeOptiona {
```
Comment (poodlewars) [inline cpp/arcticdb/version/local_versioned_engine.hpp:52]: typo - should be `SortMergeOptions`

**Review by poodlewars — CHANGES_REQUESTED** [review]

File: cpp/arcticdb/version/local_versioned_engine.hpp (hunk 6 above)
Comment (phoebusm) [reply cpp/arcticdb/version/local_versioned_engine.hpp:52]: 😿

**Review by poodlewars — APPROVED** [review]

**Pull request merged**